decisions/0014-storage-clis-for-blobstore-operations.md

```diff
@@ -0,0 +1,156 @@
+| AWS          | 🧭 Open for Contribution  |                                                                                                         |
+| GCP          | 🧭 Open for Contribution  |                                                                                                         |
+| Alibaba Cloud| 🧭 Open for Contribution  |                                                                                                         |
+
+
+## Context
+
+Cloud Controller uses the fog gem family to interface with blobstores like Azure, AWS, GCP, and Alibaba Cloud.
+These Ruby gems are largely unmaintained, introducing risks such as:
+* Dependency on deprecated SDKs (e.g., Azure SDK for Ruby)
+* Blocking Ruby version upgrades
+* Potential for unpatched CVEs
+
+Bosh faces similar issues, as it is also written in Ruby and interacts with blobstores. To address this, Bosh introduced standalone CLI tools which shell out from Ruby to handle all blobstore operations:
+- https://github.com/cloudfoundry/bosh-azure-storage-cli
+- https://github.com/cloudfoundry/bosh-s3cli
+- https://github.com/cloudfoundry/bosh-gcscli
+- https://github.com/cloudfoundry/bosh-ali-storage-cli
+
+This approach decouples core logic from Ruby gems and has proven to be robust in production.
+These CLIs are implemented in Go and use the respective provider SDKs.
+All Bosh storage CLIs implement a common interface with the following commands: `put`, `get`, `delete`, `exists`, and `sign`.
+
+A [PoC](https://github.com/cloudfoundry/cloud_controller_ng/pull/4397) has shown that `bosh-azure-storage-cli` can be successfully used in Cloud Controller to push apps.
+
+This ADR does not propose breaking changes to existing Bosh storage CLI commands or their output, but outlines necessary additions to support Cloud Controller use cases. It highlights shared concerns and encourages collaboration between Bosh and Cloud Controller.
+
+## Decision
+
+Cloud Controller will introduce support for CLI based blobstore clients, starting with Azure.
+Specifically, we will:
+* Add a new blobstore client using `bosh-azure-storage-cli`
+* Shell out from Cloud Controller to perform blobstore operations
+* Allow opt-in via `blobstore_type` configuration parameter and reusing of the existing configuration parameters:
+  * Example diff:
+    ```YAML
+      packages:
+              app_package_directory_key: app-packages
+    -          blobstore_type: storage-cli
+    +          blobstore_type: fog
+              fog_connection:
+                azure_storage_access_key: <access_key>
+                azure_storage_account_name: <account_name>
+                container_name: app-packages
+                environment: AzureCloud
+                provider: AzureRM
+              max_package_size: 1610612736
+    ```
+  * Parameters like `fog_connection` may be renamed after the transition period.       
```

Comment:
I would not define a "legacy fog_configuration" in an ADR but specify how such a configuration should really look like (of course, details can be worked out during implementation). Opt-in can easily be achieved by an ops file independent of the configuration format. Renaming bosh config parameters after a transition period is pain. Remember the experimental/temporary discussion for enabling Puma in https://github.com/cloudfoundry/capi-release/pull/483.

Questions for the configuration:
- `blobstore_type`: generic `storage-cli` or specific types per IaaS (e.g. `bosh-azure-storage-cli`)
- `connection_config` (e.g. instead of `fog_connection`): generic (unvalidated) hash or blobstore type specific configuration